### ISSUE TITLE
Discrepancy between Target Category and Actual Category Predicted by Model

### ISSUE THREAD

**Issue by tamimfaruk** [jacobgil/pytorch-grad-cam#162]
My custom CNN is tasked at predicting whether an image of movement pattern corresponds to Severe (0), Moderate (1), and Mild (2) dementias. The images are blue scribbles (of trajectories) over a white background. 

My model is able to classify with an accuracy of ~87%. 

I want to use CAM to understand with more insight which features of the movement patterns the classifier is attributing to each class of dementia. I also want to understand which features of each type of image our classifier attributes to severe dementia (i.e. CAM produced when target_category = 0). However, I noticed that there is a major discrepancy between what my model predicts for each image of my test set and the target category GradCAM produces an image for when leaving the 'target_category' parameter as None.

For example, my model might correctly classify an image as 2 (i.e. mild dementia). When inputting this same image into GradCAM, If I leave the parameter 'target category' as None, I would expect that I get the CAM corresponding to target category = 2. However, the CAM produced will be for target_category='1' (moderate dementia)

If my model is already tuned correctly, I don't understand why GradCAM is finding the highest scoring category to be contrary to what my test results show. Would appreciate hep in debugging my issue. 

Here's the code for the function I'm using:

````
def generate_CAM_image(model, imgpath, export_path=None):
  rgb_img = cv2.imread(imgpath, 1)[:, :, ::-1]
  rgb_img = cv2.resize(rgb_img, (200, 200))
  coords = get_coordinates(rgb_img)
  rgb_img = np.float32(rgb_img) / 255
  input_tensor = preprocess_image(rgb_img) #check this
  target_layer = [model.layer3[-1]]
  cam = GradCAM(model=model, target_layers=target_layer, use_cuda=True)

  # If None, returns the map for the highest scoring category.
  # Otherwise, targets the requested category.
  target_category = None

  grayscale_cam = cam(input_tensor=input_tensor,
                      target_category=target_category)

  # Here grayscale_cam has only one image in the batch
  grayscale_cam = grayscale_cam[0, :]

  cam_image = show_cam_on_image(rgb_img, grayscale_cam)
  # plt.imshow(cam_image)

  cam_image = cv2.cvtColor(cam_image, cv2.COLOR_RGB2BGR)
  # save image...
  if export_path is not None: cv2.imwrite(path, cam_image)

  cam_mean, cam_std = get_avg_pixel_intensity(grayscale_cam, coords)
  return(cam_image, cam_mean, cam_std)`
````

Here is what my model looks like.

````
class ConvNet(nn.Module):
  def __init__(self):
    super(ConvNet, self).__init__()
    self.layer1 = nn.Sequential(
      nn.Conv2d(3, 16, kernel_size=3, stride=1, padding=2),
      nn.ReLU(),
      nn.MaxPool2d(kernel_size=2, stride=2))
    self.layer2 = nn.Sequential(
      nn.Conv2d(16, 32, kernel_size=3, stride=1, padding=2),
      nn.ReLU(),
      nn.MaxPool2d(kernel_size=2, stride=2))
    self.layer3 = nn.Sequential(
      nn.Conv2d(32, 64, kernel_size=3, stride=1, padding=2),
      nn.ReLU(),
      nn.MaxPool2d(kernel_size=2, stride=2))
    self.drop_out = nn.Dropout()
    self.flatten = nn.Flatten()
    self.fc1 = nn.Linear(43264, 1000)
    self.fc2 = nn.Linear(1000, 3)

  def forward(self, x):
    out = self.layer1(x)
    out = self.layer2(out)
    out = self.layer3(out)
    out = out.reshape(out.size(0), -1)
    out = self.drop_out(out)
    out = self.flatten(out)
    out = self.fc1(out)
    out = self.fc2(out)
    return out
````

Any help and advice would be very much appreciated. Thank you.

**Comment by jacobgil** [comment]
Hi,
Sorry for the late reply.
Lets do this to debug:
- Before calling the cam algorithm, i.e before the `grayscale_cam = cam(input` line, can you please run the model on the tensor,
And print the scores?
Are these scores correct at that point?

To make sure it's correct, and there isn't, for example, some RGB-BGR problem etc.

- If it's correct, can you please modify the code to print out the value here:
`https://github.com/jacobgil/pytorch-grad-cam/blob/master/pytorch_grad_cam/base_cam.py#L75`

**Comment by tamimfaruk** [comment]
Hi Jacob,

Using your guidance, I was able to determine the error. I ran the model on my images before the line you indicated and saw that the scores were wrong even at that stage.

It turns out the issue was that I was using PIL to read my images for training, but cv2 to read my images for CAM. I'm not sure what the main differences are in the backend of each, but I revised the function above to use PIL rather than cv2. The CAMs are working as expected now. 

Thank you so much for your assistance and guidance! Your help was super appreciated.
